guardian/typerighter
# Add publish button with server-side validation

## What does this change?

Adds a publish button with server-side validation. When a rule is changed, we ping the server to ask if the rule is publishable. If it's not, we get a tooltip with a list of reasons.

![publish-validate](https://github.com/guardian/typerighter/assets/7767575/10e2dc46-3ed0-4d3e-ae28-051cd39d0696)

## How to test

- The automated tests should pass. This PR fixes a fairly egregious bug that the UI made obvious – publishing live rules affected the active status of every other live rule! Addressed in [ca0281b](https://github.com/guardian/typerighter/pull/335/commits/ca0281bed36af3c8234a8c2acc84818702ac58b8) with accompanying test.
- Try adding and publishing a few rules. It should work as expected – validation should work, and publishing valid rules should update the rule list and the rule history.

(Editing rules is a bit of a pain, now, in that when you hit 'update', you automatically navigate away from the current rule, and must re-open it to publish. Best to address in another PR, possibly by implementing real-time debounced updates.)
